Implement IRC away / unaway

vmiklos opened the pull request:
Just enough so that if your IRC client keeps an away log (like e.g.
irssi does), then it can properly start/stop as you go into away mode or
get out of it.

Fixes #73

I imagine the patch is simple enough, but if it helps, I'm happy to wait till the port to async/await is done, and then I can rebase this on top of a newer master.

vmiklos commented:
Ping. :-) Any comments other than "let's wait till the port to async/await is done"?

jplatte commented:
Hey. Seems pretty reasonable, the only thing I noticed when looking again now is that you're doing `&"foo".to_string()` which really shouldn't ever be necessary unless working with some kind of broken API. Does just the string literal not work?

vmiklos commented:
> Does just the string literal not work?

Indeed it does, fixed now. I started with format!(), then realized that I don't need any formatting and changed to just producing a String, without realizing I don't need that, either. :-)

vmiklos commented:
Ping, any idea when async_await goes in, so this can be rebased on top of that? Or is this simple enough to be accepted before the branch is merged?

jplatte commented:
Hey vmiklos. The async_await branch will probably be merged next week. You can rebase your work on top of that branch now if you want, there's most likely not going to be any changes to it before merge. I'm just waiting on the branch being tested by a few people. If you would like to help with that, please let me know whether it works for you over in [#matrix-ircd:matrix.org](https://matrix.to/#/#matrix-ircd:matrix.org).

vmiklos commented:
Thanks for the info, good to hear the merge of the branch is not far away. :-)

jplatte commented:
Just to keep you posted: Somebody found an issue with the async_await branch so it's still not merged. I notified Brooks about it, who did the async/await implementation as part of GSoC. If that doesn't result in a fix, I'll try to make time for fixing it myself in the coming weeks. Don't feel bad pinging me again if nothing has happened by the end of the month.

vmiklos commented:
Ok, thanks! :-)

vmiklos commented:
Ping, it seems the branch is still unmerged. Perhaps it would make sense to not hold this back if it requires more time for it to mature?

jplatte commented:
Sorry, the async_await branch issue still isn't fixed. I will try to get to it soon.

vmiklos commented:
Friendly ping :-)

jplatte commented:
Okay so after trying a few times I don't see the async_await port really going forward without involvement by the GSoC student who started it. This PR looks simple enough, please just [sign off](https://github.com/matrix-org/matrix-ircd/blob/master/CONTRIBUTING.md#sign-off) and I'll merge it.

vmiklos commented:
> Okay so after trying a few times I don't see the async_await port really going forward without involvement by the GSoC student who started it.

Sorry to hear that!

> please just sign off and I'll merge it.

Done.
